### DISCUSSION TITLE
A 'Keep Aspect' option for Size Flags

### DISCUSSION THREAD

**Discussion by Powerbyte7** [godotengine/godot-proposals#4039] [GUI]
If you want to have one or more nodes that expand dynamically within a container, there's no built-in option to have the margins respect an aspect ratio.

Here's a practical example of the behavior I'm looking for using a script. The VBox and the icon both maintain their aspect ratios.

https://user-images.githubusercontent.com/40431794/154808455-e4101007-aa70-44d3-85d8-af4491a6aa09.mp4

![image](https://user-images.githubusercontent.com/40431794/154809627-2b153ac0-e50e-42fa-bfaa-a3a16194a90a.png)

The way I'm achieving this behavior right now is to disable the icon's size flags and to update its minimal size instead for each frame according to the container size.
```
var size = vbox.rect_size.x/3
texture_rect.rect_min_size = Vector2(size, size)
```
I think it'd be a lot nicer to be able to set an aspect ratio of '3.0' in the VBox size flags and an aspect ratio of '1.0' in the TextureRect size flags, especially when more nodes are present in the container.

I haven't made a full proposal yet, because I'm curious to know if anyone knows a better way to do this.

**Comment by Calinou** [comment]
Did you look at the AspectRatioContainer node which was added in Godot 3.4?

**Reply by Powerbyte7** [reply]
Yes I have, the problem is specifically with margins in a container. The AspectRatioContainer helps when displaying something with a fixed aspect ratio, but it still takes up more space than necessary when using size flags to dynamically resize nodes.

![image](https://user-images.githubusercontent.com/40431794/154814322-6b1ce6a4-e700-4161-a2e7-f5cec5c9fcd0.png)
![image](https://user-images.githubusercontent.com/40431794/154814325-11598382-f6db-456f-b2b8-c8e761569a32.png)
![image](https://user-images.githubusercontent.com/40431794/154814619-f73b258e-8de8-482d-9e34-64949c44c9e9.png)

So when Godot arranges all of the nodes for you it'll try to push the margins as far as it can, but I'd like the ability to limit it to a ratio. I hope that makes sense.

**Reply by Calinou** [reply]
It sounds like https://github.com/godotengine/godot-proposals/issues/1802 could cover your use case.

**Reply by Powerbyte7** [reply]
Part of it does, thanks! 

After reading into it more I've changed my mind and I think I'll look into a script/plugin instead. It's not a major enough problem to justify complicating the layout system.
